install.packages("BiocManager")
BiocManager::install(c("DESeq2", "org.Hs.eg.db", "clusterProfiler", "RcolorBrewer", "pheatmap", "EnhancedVolcano", "colorspace", "tidyverse", 'rtracklayer'))


library('DESeq2')
library(org.Hs.eg.db)
# library(ggplot2)
library(clusterProfiler)
# library(ggrepel)
library(RColorBrewer)
library(pheatmap)
library(EnhancedVolcano)
library("colorspace") 
library(tidyverse)
# library(stringr)
library('rtracklayer')
########################


#Directory where the count data exists
dir= "C://Users/Asus/Downloads/rnaseq_workshop/rnaseq_workshop//"

#gtf file which has gene info
gtf_gene = as.data.frame(rtracklayer::import("C://Users/Asus/Downloads/rnaseq_workshop/rnaseq_workshop/Homo_sapiens.GRCh38.107.gtf")) %>%
  filter(type == "gene")
rownames(gtf_gene) = gtf_gene$gene_id

#meta file with the sample information
meta = read.table(paste0(dir,"meta.tsv"), sep = "\t", header = T)
rownames(meta) = meta$SampleID

#Reading the count file generated by featureCount
featureCount <- read.table(paste0(dir,"countMatrix.txt"), sep = "\t", header = T, check.names = F)
rownames(featureCount) = featureCount$Geneid

###################Syncing count table column name according to meta sampleID
index_array = c()
for (i in c(1:nrow(meta)))
{
  index = grep(paste0("/",meta$SampleID[i],"."), colnames(featureCount),fixed = T)
  index_array = c(index_array, index)
}
colnames(featureCount)[index_array] = as.character(meta$SampleID)
featureCount = featureCount[,which(colnames(featureCount) %in% meta$SampleID)]
meta = meta[match(colnames(featureCount), meta$SampleID),]

#Reading count data and meta table. Defining the design formula
ddsHTSeq <- DESeqDataSetFromMatrix(countData = featureCount, colData = meta, design = ~0+Condition)
ddsHTSeq

nrow(ddsHTSeq)
#Removing genes with less than 10 reads assigned across samples
ddsHTSeq <- ddsHTSeq[rowSums(counts(ddsHTSeq))>10,]
nrow(ddsHTSeq)

#regularized  log transformation
rld <- rlog(ddsHTSeq)

#Code to generate the PCA plot
pcaData <- plotPCA(rld, intgroup=c("Condition"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
color_code = c(ggsci::pal_d3(palette = "category20")(2))
ggplot(pcaData, aes(PC1, PC2, color= Condition, shape = Condition)) +
  #geom_text(aes(label=(pcaData$name),hjust=1,vjust=-1)) +
  geom_point(size=7) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  theme_bw(40) +
  #ggforce::geom_mark_ellipse(aes(fill = Condition, color = Condition), alpha = 0.1,show.legend = F ) +
  labs(title="Principal component analysis",color="Condition", shape = "Condition") 
  

#Heatmap between samples based on euclidean distance
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- colData(rld)$Condition
colnames(sampleDistMatrix) <- colData(rld)$Condition
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,main="Heatmap Euclidean distance")

#Running DESeq2
ddsHTSeq <- DESeq(ddsHTSeq, parallel=T)
resultsNames(ddsHTSeq)

#Getting the statistics for specific comparison
res.05 <- results(ddsHTSeq, alpha = 0.05, contrast = c(-1,1))
#res.05 <- results(ddsHTSeq, alpha = 0.05, contrast = c("Condition","Treatment", "Control"))

#adding the gene symbol, gene name and gene biotype to the results of DESeq2
res.05$symbol <- mapIds(org.Hs.eg.db,
                        keys=row.names(res.05),
                        column="SYMBOL",
                        keytype="ENSEMBL",
                        multiVals="first")
res.05$genename <- mapIds(org.Hs.eg.db,
                          keys=row.names(res.05),
                          column="GENENAME",
                          keytype="ENSEMBL",
                          multiVals="first")
res.05$entrez <- mapIds(org.Hs.eg.db,
                          keys=row.names(res.05),
                          column="ENTREZID",
                          keytype="ENSEMBL",
                          multiVals="first")
res.05$gene_biotype = gtf_gene[row.names(res.05),]$gene_biotype
write.table(res.05, file = paste0(dir,"DEG_table.tsv"), sep = "\t", row.names = T, col.names = NA)

resSig <- subset(res.05, padj < 0.05)
summary(resSig)
resSig <- subset(resSig, abs(log2FoldChange) > 1)
summary(resSig)

#Choosing differentially expressed genes based on log2 fold change cutoff
resSigUp <- subset(resSig, (log2FoldChange) > 1)
resSigDown <- subset(resSig, (log2FoldChange) < -1)
write.table(resSigUp, file = paste0(dir,"/",foldchangeFolder,"/DEG_Sig_Up.tsv"), sep = "\t", row.names = T, col.names = NA)
write.table(resSigDown, file = paste0(dir,"/",foldchangeFolder,"/DEG_Sig_Down.tsv"), sep = "\t", row.names = T, col.names = NA)


#Code for generating heatmap of the differentially expressed up and down regulated genes
my_sample_col <- data.frame(colData(rld))

assayrld = assay(rld) %>% as.data.frame()
assayrld = rbind(assayrld[which((rownames(assayrld) %in% rownames(resSigUp))),],
                 assayrld[which((rownames(assayrld) %in% rownames(resSigDown))),])

my_sample_row <- data.frame(c(rep("Up-regulated", nrow(resSigUp)),
                              rep("Down-regulated", nrow(resSigDown))))

row.names(my_sample_row) <- rownames(assayrld)
colnames(my_sample_row) = "Differential Genes"

pheatmap(assayrld, cluster_rows=T, show_rownames=F, show_colnames = T,
         cluster_cols=T, annotation_col= tibble(my_sample_col) %>% column_to_rownames('SampleID'), annotation_row = my_sample_row,
         scale = "row", color = colorRampPalette(c("navy", "white", "firebrick3"))(100),
         clustering_distance_cols = "euclidean") +
  theme_bw(22)

#Code to generate volcano plot
res.05$pathway = NA
res.05$pathway[which((res.05$padj < 0.05) & (res.05$log2FoldChange > 1) & (is.na(res.05$pathway)))] = "Significant Up"
res.05$pathway[which((res.05$padj < 0.05) & (res.05$log2FoldChange < -1) & (is.na(res.05$pathway)))] = "Significant Down"

keyvals <- ifelse(res.05$pathway == "Significant Up", 'blue',
                  ifelse(res.05$pathway == "Significant Down", 'red',
                         'grey90'))
keyvals[is.na(keyvals)] <- 'grey90'
names(keyvals)[keyvals == 'blue'] <- 'Significantly up-regulated'
names(keyvals)[keyvals == 'red'] <- 'Significantly down-regulated'
names(keyvals)[keyvals == 'grey90'] <- 'Not significant'

EnhancedVolcano(res.05,
                lab = res.05$symbol,
                #lab = NA,  
                x = 'log2FoldChange',
                y = 'padj',
                ylab = "-log10(padj)",
                xlab = "log2FC",
                title = "Differentially expressed genes",
                FCcutoff = 1,
                pCutoff = 0.05,
                legendPosition = 'right',
                legendLabSize = 16,
                legendIconSize = 5.0,
                colCustom = keyvals,
                subtitle = ""
                #pointSize = c(ifelse(res$log2FoldChange>2, 8, 1))
)+
  theme(plot.title = element_text(hjust = 0.5))

#Selecting the protein coding genes for the functional enrichment analysis
resSigUp_PC =  subset(resSigUp, gene_biotype == "protein_coding")
resSigDown_PC =  subset(resSigDown, gene_biotype == "protein_coding")
write.table(resSigUp_PC, file = paste0(outFolder,"/",foldchangeFolder,"/DEG_Sig_Up_PC.tsv"), sep = "\t", row.names = T, col.names = NA)
write.table(resSigDown_PC, file = paste0(outFolder,"/",foldchangeFolder,"/DEG_Sig_Down_PC.tsv"), sep = "\t", row.names = T, col.names = NA)

#GO Enrichment
EGO_up <- enrichGO(gene = rownames(resSigUp_PC),keyType = 'ENSEMBL', OrgDb = org.Hs.eg.db, ont = "BP")
simpEGO_up = clusterProfiler::simplify(EGO_up, cutoff = 0.7, by = "p.adjust", select_fun = min, measure = "Wang")
dotplot(simpEGO_up, orderBy = "p.adjust")

EGO_dn <- enrichGO(gene = rownames(resSigDown_PC),keyType = 'ENSEMBL', OrgDb = org.Hs.eg.db, ont = "BP")
simpEGO_dn = clusterProfiler::simplify(EGO_dn, cutoff = 0.7, by = "p.adjust", select_fun = min, measure = "Wang")
dotplot(simpEGO_dn, orderBy = "p.adjust")

#KEGG Enrichment
EKEGG_up = enrichKEGG(gene = resSigUp_PC$entrez, organism = "hsa", pvalueCutoff = 0.05)
dotplot(EKEGG_up, orderBy = "p.adjust")

EKEGG_dn = enrichKEGG(gene = resSigDown_PC$entrez, organism = "hsa", pvalueCutoff = 0.05)
dotplot(EKEGG_dn, orderBy = "p.adjust")
